proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,393 @@
+
+# Background
+
+Cryptography is an essential part of modern computing and information security, but it requires care, nuance, and knowledge to properly implement, and is extremely easy to mess up. It covers a diverse range of functions, with individual algorithms often having their own particular quirks that must be acknowledged to properly and safely use. Failure to properly implement cryptography carries the risk of severely compromising the security of any system in which it is used. Manually implementing common cryptographic primitives such as hashes, ciphers, digital signatures, big-integer arithmetic is thus considered extremely unwise, both in terms of security and performance.
+
+Cryptography is even more difficult for functional programming languages. Properly implementing techniques such as sanitizing memory, performing operations in constant time, and zeroing and ejecting from caches / registers, are already a challenge in a strict imperative language such as C, and in a lazy functional language it is even more difficult. Lazy evaluation presents a challenge towards managing and clearing sensitive data from memory, as care must be taken in order to ensure that it is not captured as part of a closure or thunk, and to force finalizers / cleanup appropriately or else the data may stay in memory for longer than intended. As a result, the space of functional cryptography appears to be relatively under-explored and under-developed, having favored other languages where such things are easier to manage.
+
+However, if we accept this downside as a part of a trade-off, and explore the ways that functional programming is beneficial to cryptography, there is then also the potential for functional cryptography to be made *easier*.
+
+The collective properties of referential transparency, type systems, and monadic interfaces provide a safer and more lawful context in which to write functions than the free-for-all that many imperative / OOP languages can offer, and so the functional machinery of Haskell is well-suited for adding important measures of safety and control to cryptography, helping to prevent or eliminate many issues and errors in ways that languages such as C cannot. I believe that there is room for significant improvement, given a directed effort to develop the space.
+
+# Problem Statement
+
+The Haskell cryptography ecosystem lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and it exposes end users to significant risk in the event of a lapse in security. There are commercial solutions, but there is no community-driven, community-owned solution, and one is unlikely to be developed without deliberate community effort (see Appendix A - Commercial Incentive Misalignment).
```

Comment:
Which commercial solutions do you have in mind? It would be nice to link such prior art if any.

Reply:
I was mostly thinking of Cardano, but also that there are commercial cryptography libraries that could probably be paid to develop Haskell bindings for their proprietary code. An alternative wording might be "There are ways to solve this commercially".

Reply:
My impression of Haskell cryptography in industry is that the majority of companies just use `cryptonite` / `crypton` or other preexisting packages. Is Cardano any different? Could they possibly chime in and share their experience? @michaelpj @angerman

Reply:
Cardano is all open source, the library of ours that contains crypto bindings is Apache licensed. But it _is_ very geared towards our usage, e.g. it contains some bindings to functions that only exist in our fork of libsodium 🙈 So I don't think our stuff is a "commercial solution" in that it's not commercial and it's probably also not a solution to your problems!

I might have the history wrong, but I think we used to use cryptonite but now just write our own bindings to relevant C libraries. I don't know why we stopped using cryptonite, but probably some combination of worries about maintainership and it significantly increasing our audit surface.